pkg/lib/cockpit-components-typeahead-select.tsx

```diff
@@ -0,0 +1,428 @@
+
+/* This is a copy of TypeaheadSelect.tsx from
+
+       https://github.com/patternfly/patternfly-react/blob/v5/packages/react-templates/src/components/Select/TypeaheadSelect.tsx
+
+   We don't use it directly from the @patternfly/react-templates node
+   module since we want to add features to it, and also to isolate us
+   from gratuitous upstream changes.
+*/
+
+import React from 'react';
+import {
+  Select,
+  SelectOption,
+  SelectList,
+  SelectOptionProps,
+  MenuToggle,
+  MenuToggleElement,
+  TextInputGroup,
+  TextInputGroupMain,
+  TextInputGroupUtilities,
+  Button,
+  MenuToggleProps,
+  SelectProps
+} from '@patternfly/react-core';
+import TimesIcon from '@patternfly/react-icons/dist/esm/icons/times-icon';
+
+export interface TypeaheadSelectOption extends Omit<SelectOptionProps, 'content' | 'isSelected'> {
+  /** Content of the select option. */
+  content: string | number;
+  /** Value of the select option. */
+  value: string | number;
+  /** Indicator for option being selected */
+  isSelected?: boolean;
+}
+
+export interface TypeaheadSelectProps extends Omit<SelectProps, 'toggle' | 'onSelect'> {
+  /** @hide Forwarded ref */
+  innerRef?: React.Ref<any>;
+  /** Options of the select */
+  selectOptions: TypeaheadSelectOption[];
+  /** Callback triggered on selection. */
+  onSelect?: (
+    _event: React.MouseEvent<Element, MouseEvent> | React.KeyboardEvent<HTMLInputElement> | undefined,
+    selection: string | number
+  ) => void;
+  /** Callback triggered when the select opens or closes. */
+  onToggle?: (nextIsOpen: boolean) => void;
+  /** Callback triggered when the text in the input field changes. */
+  onInputChange?: (newValue: string) => void;
+  /** Function to return items matching the current filter value */
+  filterFunction?: (filterValue: string, options: TypeaheadSelectOption[]) => TypeaheadSelectOption[];
+  /** Callback triggered when the clear button is selected */
+  onClearSelection?: () => void;
+  /** Placeholder text for the select input. */
+  placeholder?: string;
+  /** Flag to indicate if the typeahead select allows new items */
+  isCreatable?: boolean;
+  /** Flag to indicate if create option should be at top of typeahead */
+  isCreateOptionOnTop?: boolean;
+  /** Message to display to create a new option */
+  createOptionMessage?: string | ((newValue: string) => string);
+  /** Message to display when no options are available. */
+  noOptionsAvailableMessage?: string;
+  /** Message to display when no options match the filter. */
+  noOptionsFoundMessage?: string | ((filter: string) => string);
+  /** Flag indicating the select should be disabled. */
+  isDisabled?: boolean;
+  /** Width of the toggle. */
+  toggleWidth?: string;
+  /** Additional props passed to the toggle. */
+  toggleProps?: MenuToggleProps;
+}
+
+const defaultNoOptionsFoundMessage = (filter: string) => `No results found for "${filter}"`;
+const defaultCreateOptionMessage = (newValue: string) => `Create "${newValue}"`;
```

Comment:
Indeed! Does PF itself not i18n their strings in general?

Reply:
I'm not aware of a mechanism to even tell PF components about the current language. I suppose we should just make the corresponding attributes mandatory, i.e. always provide the text from the caller?

Reply:
Yes, that's better.

Reply:
I have changed my mind. It would be tedious to force all users to provide a generic "Not found" string. I have done the l10n in TypeaheadSelect, and used strings that are already translated.